From 159850628cbe6b89379e93b5e4cde714637a8c09 Mon Sep 17 00:00:00 2001
From: locus84 <locus84@andestech.com>
Date: Mon, 6 Jan 2025 16:04:24 +0800
Subject: [PATCH 33/53] andes: spi: SPI support data merge mode (#142)

Currently, SPI Supports data merge mode, each time the data is written,
the register will send all four bytes of the written data.

Each read from the Data Register will retrieve four bytes
of received data as a single word data.

Signed-off-by: locus84 <locus84@andestech.com>
Reviewed-on: https://gitea.andestech.com/RD-SW/u-boot.official/pulls/142
Reviewed-by: Tim Shih-Ting OuYang <tim609@andestech.com>
Reviewed-by: CL Chin-Long Wang <cl634@andestech.com>
Co-authored-by: locus84 <locus84@andestech.com>
Co-committed-by: locus84 <locus84@andestech.com>
---
 configs/ae350_rv32_defconfig            |   1 +
 configs/ae350_rv32_falcon_defconfig     |   1 +
 configs/ae350_rv32_falcon_xip_defconfig |   1 +
 configs/ae350_rv32_spl_defconfig        |   1 +
 configs/ae350_rv32_spl_xip_defconfig    |   1 +
 configs/ae350_rv32_xip_defconfig        |   1 +
 configs/ae350_rv64_defconfig            |   1 +
 configs/ae350_rv64_falcon_defconfig     |   1 +
 configs/ae350_rv64_falcon_xip_defconfig |   1 +
 configs/ae350_rv64_spl_defconfig        |   1 +
 configs/ae350_rv64_spl_xip_defconfig    |   1 +
 configs/ae350_rv64_xip_defconfig        |   1 +
 drivers/spi/Kconfig                     |   7 +
 drivers/spi/atcspi200_spi.c             | 232 +++++++++++++-----------
 14 files changed, 142 insertions(+), 109 deletions(-)

diff --git a/configs/ae350_rv32_defconfig b/configs/ae350_rv32_defconfig
index d2065d399ed0..24e1dbf4e312 100644
--- a/configs/ae350_rv32_defconfig
+++ b/configs/ae350_rv32_defconfig
@@ -50,6 +50,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_WATCHDOG_AUTOSTART is not set
 CONFIG_WDT=y
 CONFIG_WDT_ATCWDT200=y
diff --git a/configs/ae350_rv32_falcon_defconfig b/configs/ae350_rv32_falcon_defconfig
index 6cd2e11aea34..80d7c262d8f7 100644
--- a/configs/ae350_rv32_falcon_defconfig
+++ b/configs/ae350_rv32_falcon_defconfig
@@ -59,6 +59,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_BINMAN_FDT is not set
 CONFIG_CONSOLE_RECORD=y
 CONFIG_CONSOLE_RECORD_OUT_SIZE_F=0x2000
diff --git a/configs/ae350_rv32_falcon_xip_defconfig b/configs/ae350_rv32_falcon_xip_defconfig
index 91a8d0c32bdf..425ffc2c14c2 100644
--- a/configs/ae350_rv32_falcon_xip_defconfig
+++ b/configs/ae350_rv32_falcon_xip_defconfig
@@ -60,6 +60,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_BINMAN_FDT is not set
 CONFIG_CONSOLE_RECORD=y
 CONFIG_CONSOLE_RECORD_OUT_SIZE_F=0x2000
diff --git a/configs/ae350_rv32_spl_defconfig b/configs/ae350_rv32_spl_defconfig
index 6c16dd275036..5b58213047ff 100644
--- a/configs/ae350_rv32_spl_defconfig
+++ b/configs/ae350_rv32_spl_defconfig
@@ -60,6 +60,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_WATCHDOG_AUTOSTART is not set
 CONFIG_WDT=y
 CONFIG_WDT_ATCWDT200=y
diff --git a/configs/ae350_rv32_spl_xip_defconfig b/configs/ae350_rv32_spl_xip_defconfig
index 88614b1fbbe0..3c018cde2d53 100644
--- a/configs/ae350_rv32_spl_xip_defconfig
+++ b/configs/ae350_rv32_spl_xip_defconfig
@@ -61,6 +61,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_WATCHDOG_AUTOSTART is not set
 CONFIG_WDT=y
 CONFIG_WDT_ATCWDT200=y
diff --git a/configs/ae350_rv32_xip_defconfig b/configs/ae350_rv32_xip_defconfig
index 2d62107199d0..368d2ffd3fe4 100644
--- a/configs/ae350_rv32_xip_defconfig
+++ b/configs/ae350_rv32_xip_defconfig
@@ -51,6 +51,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_WATCHDOG_AUTOSTART is not set
 CONFIG_WDT=y
 CONFIG_WDT_ATCWDT200=y
diff --git a/configs/ae350_rv64_defconfig b/configs/ae350_rv64_defconfig
index a583fc317b7d..6f0cc6bc1d06 100644
--- a/configs/ae350_rv64_defconfig
+++ b/configs/ae350_rv64_defconfig
@@ -50,6 +50,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_WATCHDOG_AUTOSTART is not set
 CONFIG_WDT=y
 CONFIG_WDT_ATCWDT200=y
diff --git a/configs/ae350_rv64_falcon_defconfig b/configs/ae350_rv64_falcon_defconfig
index b53c9febf48c..a623e3b2b5fd 100644
--- a/configs/ae350_rv64_falcon_defconfig
+++ b/configs/ae350_rv64_falcon_defconfig
@@ -59,6 +59,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_BINMAN_FDT is not set
 CONFIG_CONSOLE_RECORD=y
 CONFIG_CONSOLE_RECORD_OUT_SIZE_F=0x2000
diff --git a/configs/ae350_rv64_falcon_xip_defconfig b/configs/ae350_rv64_falcon_xip_defconfig
index 92d73acf3b5d..4de6e2e6a52e 100644
--- a/configs/ae350_rv64_falcon_xip_defconfig
+++ b/configs/ae350_rv64_falcon_xip_defconfig
@@ -60,6 +60,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_BINMAN_FDT is not set
 CONFIG_CONSOLE_RECORD=y
 CONFIG_CONSOLE_RECORD_OUT_SIZE_F=0x2000
diff --git a/configs/ae350_rv64_spl_defconfig b/configs/ae350_rv64_spl_defconfig
index ceb092f8d7fd..45ee30fc668f 100644
--- a/configs/ae350_rv64_spl_defconfig
+++ b/configs/ae350_rv64_spl_defconfig
@@ -60,6 +60,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_WATCHDOG_AUTOSTART is not set
 CONFIG_WDT=y
 CONFIG_WDT_ATCWDT200=y
diff --git a/configs/ae350_rv64_spl_xip_defconfig b/configs/ae350_rv64_spl_xip_defconfig
index 517dc5474187..ce2952a9670d 100644
--- a/configs/ae350_rv64_spl_xip_defconfig
+++ b/configs/ae350_rv64_spl_xip_defconfig
@@ -61,6 +61,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_WATCHDOG_AUTOSTART is not set
 CONFIG_WDT=y
 CONFIG_WDT_ATCWDT200=y
diff --git a/configs/ae350_rv64_xip_defconfig b/configs/ae350_rv64_xip_defconfig
index 7adb1298e838..caf4fe4a7654 100644
--- a/configs/ae350_rv64_xip_defconfig
+++ b/configs/ae350_rv64_xip_defconfig
@@ -51,6 +51,7 @@ CONFIG_BAUDRATE=115200
 CONFIG_SYS_NS16550=y
 CONFIG_SPI=y
 CONFIG_ATCSPI200_SPI=y
+CONFIG_ATCSPI200_SPI_DATA_MERGE=y
 # CONFIG_WATCHDOG_AUTOSTART is not set
 CONFIG_WDT=y
 CONFIG_WDT_ATCWDT200=y
diff --git a/drivers/spi/Kconfig b/drivers/spi/Kconfig
index 35030ab35561..d8e2d2fe988f 100644
--- a/drivers/spi/Kconfig
+++ b/drivers/spi/Kconfig
@@ -74,6 +74,13 @@ config ATCSPI200_SPI
 	  used to access the SPI flash on AE3XX and AE250 platforms embedding
 	  this Andestech IP core.
 
+config ATCSPI200_SPI_DATA_MERGE
+	bool "Enable ATCSPI200 data merge mode"
+	help
+	  This option enables data merge mode for the ATCSPI200 driver.
+	  When enabled, each write to the data register transmits four bytes,
+	  and each read retrieves four bytes as a single word.
+
 config ATH79_SPI
 	bool "Atheros SPI driver"
 	depends on ARCH_ATH79
diff --git a/drivers/spi/atcspi200_spi.c b/drivers/spi/atcspi200_spi.c
index 154a14646ef7..483388dcbd74 100644
--- a/drivers/spi/atcspi200_spi.c
+++ b/drivers/spi/atcspi200_spi.c
@@ -31,11 +31,19 @@ DECLARE_GLOBAL_DATA_PTR;
 #define SPI_DEF_SRC_CLK		100000000
 #define SPI_DEF_MAX_CLK		40000000
 
-#define DATA_LENGTH(x)		((x - 1) << 8)
-#define ADDR_LENGTH(x)		((x - 1) << 16)
+/* SPI Transfer Format Register */
 #define ADDR_LEN_OFFSET		16
+#define ADDR_BIT		0x3
+#define ADDR_LENGTH(x)		((x - 1) << ADDR_LEN_OFFSET)
 #define ADDR_LEN_MASK		(0x3 << ADDR_LEN_OFFSET)
-
+#define DATA_LENGTH_OFFSET	8
+#define DATA_BIT		0x8
+#define DATA_LENGTH(x)		((x - 1) << DATA_LENGTH_OFFSET)
+#define DATA_LENGTH_MASK	(0x1f << DATA_LENGTH_OFFSET)
+#define DATA_MERGE_OFFSET	7
+#define DATA_MERGE		(1 << DATA_MERGE_OFFSET)
+
+/* SPI Transfer Control Register */
 #define TRAMODE_OFFSET		24
 #define TRAMODE_MASK		(0x0F << TRAMODE_OFFSET)
 #define TRAMODE_WR_SYNC		(0 << TRAMODE_OFFSET)
@@ -54,21 +62,20 @@ DECLARE_GLOBAL_DATA_PTR;
 #define RCNT_OFFSET		0
 #define RCNT_MASK		(CNT_MASK << RCNT_OFFSET)
 
-#define TXFTH_OFFSET		16
-#define RXFTH_OFFSET		8
-#define TXDMAEN			(1 << 4)
-#define RXDMAEN			(1 << 3)
-#define TXFRST			(1 << 2)
-#define RXFRST			(1 << 1)
-#define SPIRST			(1 << 0)
-
+/* SPI Status Register */
+#define TXFVE_OFFSET		16
+#define RXFVE_OFFSET		8
 #define TXFFL			(1 << 23)
 #define TXEPTY			(1 << 22)
 #define TXFVE_MASK		(0x1F << 16)
-#define TXFVE_OFFSET		16
 #define RXFEM			(1 << 14)
-#define RXFVE_OFFSET		8
 #define RXFVE_MASK		(0x1F << RXFVE_OFFSET)
+
+#define TXDMAEN			(1 << 4)
+#define RXDMAEN			(1 << 3)
+#define TXFRST			(1 << 2)
+#define RXFRST			(1 << 1)
+#define SPIRST			(1 << 0)
 #define SPIBSY			(1 << 0)
 
 #define SCLK_DIV_MASK		0xFF
@@ -108,8 +115,9 @@ struct nds_spi_subordinate {
 	u8		cmd_buf[16];
 	size_t		data_len;
 	size_t		tran_len;
-	u8		*din;
-	u8		*dout;
+	void		*din;
+	const void	*dout;
+	bool		data_merge;
 	unsigned int    max_transfer_length;
 };
 
@@ -172,7 +180,17 @@ static int __atcspi200_spi_claim_bus(struct nds_spi_subordinate *ns)
 			return -EINVAL;
 
 	ns->cmd_len = 0;
-	format = ns->mode | DATA_LENGTH(8) | ADDR_LENGTH(3);
+#ifdef	CONFIG_ATCSPI200_SPI_DATA_MERGE
+	format = ns->mode | DATA_LENGTH(DATA_BIT) | ADDR_LENGTH(ADDR_BIT) | DATA_MERGE;
+	ns->data_merge = true;
+	if ((format & DATA_LENGTH_MASK) >> DATA_LENGTH_OFFSET != 0x7) {
+		ns->data_merge = false;
+		pr_warn("Warning: DATA merge mode takes effect when DataLen = 0x7\n");
+	}
+#else
+	format = ns->mode | DATA_LENGTH(DATA_BIT) | ADDR_LENGTH(ADDR_BIT);
+	ns->data_merge = false;
+#endif
 	ns->regs->format = format;
 	ret = __atcspi200_spi_set_speed(ns);
 
@@ -215,33 +233,94 @@ static int __atcspi200_spi_stop(struct nds_spi_subordinate *ns)
 	return 0;
 }
 
-static void __nspi_espi_tx(struct nds_spi_subordinate *ns, const void *dout)
+static inline void __nspi_espi_tx(struct nds_spi_subordinate *ns, const void *dout)
 {
-	ns->regs->data = *(u8 *)dout;
+	if (ns->data_merge)
+		ns->regs->data = *(u32 *)dout;
+	else
+		ns->regs->data = *(u8 *)dout;
 }
 
-static int __nspi_espi_rx(struct nds_spi_subordinate *ns, void *din, unsigned int bytes)
+static inline void __nspi_espi_rx(struct nds_spi_subordinate *ns, void *din)
 {
-	*(u8 *)din = ns->regs->data;
-	return bytes;
+	if (ns->data_merge)
+		*(u32 *)din = ns->regs->data;
+	else
+		*(u8 *)din = ns->regs->data;
 }
 
+static void pio_transfer_data(struct nds_spi_subordinate *ns,
+			      void *data_in, const void *data_out, int num_blks)
+{
+	int tx_count = 0, rx_count = 0;
+	int num_byte;
+	int timeout;
+
+	if (ns->dout)
+		tx_count = ((ns->regs->tctrl >> 12) & CNT_MASK) + 1;
+	if (ns->din)
+		rx_count = ((ns->regs->tctrl >> 0) & CNT_MASK) + 1;
+
+	num_byte = (ns->regs->format & DATA_MERGE) ? 4 : 1;
+	timeout = SPI_TIMEOUT;
+	while (num_blks && timeout--) {
+		unsigned int event;
+		unsigned int transfer_bytes;
+		int i;
+		int xf_cnt;
+
+		event = in_le32(&ns->regs->status);
+		if (!(event & TXFFL) && data_out) {
+			xf_cnt = (TXFVE_MASK - (event & TXFVE_MASK)) >> TXFVE_OFFSET;
+			transfer_bytes = min3(xf_cnt, CHUNK_SIZE, num_blks);
+			for (i = 0; i < transfer_bytes; i++) {
+				__nspi_espi_tx(ns, data_out);
+				data_out = (u8 *)data_out + num_byte;
+				tx_count = tx_count - num_byte;
+			}
+		}
+		event = in_le32(&ns->regs->status);
+		if ((event & RXFVE_MASK) && data_in) {
+			xf_cnt = (event & RXFVE_MASK) >> RXFVE_OFFSET;
+			transfer_bytes = min3(xf_cnt, CHUNK_SIZE, num_blks);
+			for (i = 0; i < transfer_bytes; i++) {
+				__nspi_espi_rx(ns, data_in);
+				data_in = (u8 *)data_in + num_byte;
+				rx_count = rx_count - num_byte;
+			}
+		}
+		num_blks = max(tx_count, rx_count);
+
+		if (!timeout) {
+			debug("spi_xfer: %s() timeout\n", __func__);
+			break;
+		}
+	}
+}
 
 static int __atcspi200_spi_xfer(struct nds_spi_subordinate *ns, unsigned int bitlen,
 		const void *data_out, void *data_in, unsigned long flags)
 {
-	unsigned int event, rx_bytes;
-	const void *dout = NULL;
-	void *din = NULL;
 	int num_blks, num_chunks, max_tran_len, tran_len;
 	int num_bytes;
 	u8 *cmd_buf = ns->cmd_buf;
 	size_t cmd_len = ns->cmd_len;
 	unsigned long data_len = bitlen / 8;
-	int rf_cnt;
 	int ret = 0, timeout = 0;
-	int tx_count = 0, rx_count = 0;
+	int format_val;
 
+	format_val = ns->regs->format;
+	if (IS_ENABLED(CONFIG_ATCSPI200_SPI_DATA_MERGE)) {
+		if (data_len % 4) {
+			format_val = format_val & ~DATA_MERGE;
+			ns->regs->format = format_val;
+			ns->data_merge = false;
+		} else {
+			format_val = format_val | DATA_MERGE;
+			ns->regs->format = format_val;
+			ns->data_merge = true;
+		}
+	}
 	max_tran_len = ns->max_transfer_length;
 	switch (flags) {
 	case SPI_XFER_BEGIN:
@@ -255,14 +334,14 @@ static int __atcspi200_spi_xfer(struct nds_spi_subordinate *ns, unsigned int bit
 			return 0;
 		}
 		ns->data_len = data_len;
-		ns->din = (u8 *)data_in;
-		ns->dout = (u8 *)data_out;
+		ns->din = data_in;
+		ns->dout = data_out;
 		break;
 
 	case SPI_XFER_BEGIN | SPI_XFER_END:
 		ns->data_len = data_len;
-		ns->din = (u8 *)data_in;
-		ns->dout = (u8 *)data_out;
+		ns->din = data_in;
+		ns->dout = data_out;
 		break;
 	default:
 		BUG_ON(1);
@@ -273,8 +352,6 @@ static int __atcspi200_spi_xfer(struct nds_spi_subordinate *ns, unsigned int bit
 		      *(uint *)data_out, data_out, *(uint *)data_in,
 		      data_in, data_len);
 	num_chunks = DIV_ROUND_UP(data_len, max_tran_len);
-	din = data_in;
-	dout = data_out;
 	while (num_chunks--) {
 		tran_len = min((size_t)data_len, (size_t)max_tran_len);
 		ns->tran_len = tran_len;
@@ -284,42 +361,7 @@ static int __atcspi200_spi_xfer(struct nds_spi_subordinate *ns, unsigned int bit
 		if(num_bytes == 0)
 			num_bytes = CHUNK_SIZE;
 		__atcspi200_spi_start(ns);
-		if (ns->dout)
-			tx_count = ((ns->regs->tctrl >> 12) & CNT_MASK) + 1;
-		if (ns->din)
-			rx_count = ((ns->regs->tctrl >> 0) & CNT_MASK) + 1;
-
-		while (num_blks && (timeout--)) {
-			event = in_le32(&ns->regs->status);
-			if ((event & TXEPTY) && data_out && tx_count) {
-				__nspi_espi_tx(ns, dout);
-				dout += CHUNK_SIZE;
-				tx_count--;
-			}
-
-			event = in_le32(&ns->regs->status);
-			if ((event & RXFVE_MASK) && data_in && rx_count) {
-				rf_cnt = ((event & RXFVE_MASK)>> RXFVE_OFFSET);
-				if (rf_cnt >= CHUNK_SIZE)
-					rx_bytes = CHUNK_SIZE;
-				else if (num_blks == 1 && rf_cnt == num_bytes)
-					rx_bytes = num_bytes;
-				else
-					continue;
-
-				if (__nspi_espi_rx(ns, din, rx_bytes) == rx_bytes) {
-					din = (unsigned char *)din + rx_bytes;
-					rx_count--;
-				}
-			}
-
-			if (!timeout) {
-				debug("spi_xfer: %s() timeout\n", __func__);
-				break;
-			}
-			num_blks = max(tx_count, rx_count);
-		}
-
+		pio_transfer_data(ns, ns->din, ns->dout, num_blks);
 		data_len -= tran_len;
 		if (data_len) {
 			ns->cmd_buf[1] += (tran_len >> 16) & 0xFF;
@@ -334,47 +376,8 @@ static int __atcspi200_spi_xfer(struct nds_spi_subordinate *ns, unsigned int bit
 	return ret;
 }
 
-static void pio_transfer_data(struct nds_spi_subordinate *ns,
-			      u8 *data_in, u8 *data_out, int num_blks)
-{
-	u8 *dout = data_out;
-	u8 *din = data_in;
-	unsigned int event, transfer_bytes;
-	int timeout, rf_cnt, tf_cnt, i;
-
-	timeout = SPI_TIMEOUT;
-	while (num_blks && (timeout--)) {
-		event = in_le32(&ns->regs->status);
-
-		if (!(event & TXFFL) && (data_out)) {
-			tf_cnt = ((TXFVE_MASK - (event & TXFVE_MASK)) >> TXFVE_OFFSET);
-			transfer_bytes = min3(tf_cnt, CHUNK_SIZE, num_blks);
-			for (i = 0; i < transfer_bytes; i++) {
-				__nspi_espi_tx(ns, dout);
-				dout += 1;
-			}
-			num_blks -= transfer_bytes;
-		}
-
-		if ((event & RXFVE_MASK) && (data_in)) {
-			rf_cnt = ((event & RXFVE_MASK) >> RXFVE_OFFSET);
-			transfer_bytes = min3(rf_cnt, CHUNK_SIZE, num_blks);
-			for (i = 0; i < transfer_bytes; i++) {
-				__nspi_espi_rx(ns, din, transfer_bytes);
-				din = (unsigned char *)din + 1;
-			}
-			num_blks -= transfer_bytes;
-		}
-
-		if (!timeout) {
-			debug("spi_xfer: %s() timeout\n", __func__);
-			break;
-		}
-	}
-}
-
 static void dma_transfer_data(struct nds_spi_subordinate *ns,
-			      u8 *data_in, u8 *data_out, int num_blks)
+			      void *data_in, const void *data_out, int num_blks)
 {
 	/* Not implemented yet */
 	pr_err("DMA transfer mode is not supported.\n");
@@ -537,11 +540,22 @@ static int atcspi200_spi_mem_exec_op(struct spi_slave *subordinate,
 	ns = dev_get_priv(bus);
 	format_val = ns->regs->format;
 	if (op->data.nbytes) {
+		if (IS_ENABLED(CONFIG_ATCSPI200_SPI_DATA_MERGE)) {
+			if (op->data.nbytes % 4) {
+				format_val = format_val & (~DATA_MERGE);
+				ns->regs->format = format_val;
+				ns->data_merge = false;
+			} else {
+				format_val = format_val | DATA_MERGE;
+				ns->regs->format = format_val;
+				ns->data_merge = true;
+			}
+		}
 		if (op->data.dir == SPI_MEM_DATA_IN) {
 			ns->din = op->data.buf.in;
 			ns->dout = 0;
 		} else {
-			ns->dout = (char *)op->data.buf.out;
+			ns->dout = op->data.buf.out;
 			ns->din = 0;
 		}
 	}
-- 
2.34.1

